Sources/GRPCHTTP2Core/GRPCStreamStateMachine.swift

```diff
@@ -0,0 +1,1391 @@
+        // Respond with HTTP-level Unsupported Media Type status code.
+        var trailers = HPACKHeaders()
+        trailers.add("415", forKey: .status)
+        return .rejectRPC(trailers: trailers)
+      }
+
+      let path = metadata.firstString(forKey: .path)
+        .flatMap { MethodDescriptor(fullyQualifiedMethod: $0) }
+      guard path != nil else {
+        let status = Status(
+          code: .unimplemented,
+          message: "No \(GRPCHTTP2Keys.path.rawValue) header has been set."
+        )
+        let trailers = self.makeTrailers(status: status, customMetadata: nil, trailersOnly: true)
+        return .rejectRPC(trailers: trailers)
+      }
+
+      func isIdentityOrCompatibleEncoding(_ clientEncoding: CompressionAlgorithm) -> Bool {
+        clientEncoding == .identity || configuration.acceptedEncodings.contains(clientEncoding)
+      }
+
+      // Firstly, find out if we support the client's chosen encoding, and reject
+      // the RPC if we don't.
+      let inboundEncoding: CompressionAlgorithm
+      let encodingValues = metadata.values(
+        forHeader: GRPCHTTP2Keys.encoding.rawValue,
+        canonicalForm: true
+      )
+      var encodingValuesIterator = encodingValues.makeIterator()
+      if let rawEncoding = encodingValuesIterator.next() {
+        guard encodingValuesIterator.next() == nil else {
+          let status = Status(
+            code: .internalError,
+            message: "\(GRPCHTTP2Keys.encoding) must contain no more than one value."
+          )
+          let trailers = self.makeTrailers(status: status, customMetadata: nil, trailersOnly: true)
+          return .rejectRPC(trailers: trailers)
+        }
+
+        guard let clientEncoding = CompressionAlgorithm(rawValue: String(rawEncoding)),
+          isIdentityOrCompatibleEncoding(clientEncoding)
+        else {
+          if configuration.acceptedEncodings.isEmpty {
+            let status = Status(code: .unimplemented, message: "Compression is not supported")
+            let trailers = self.makeTrailers(
+              status: status,
+              customMetadata: nil,
+              trailersOnly: true
+            )
+            return .rejectRPC(trailers: trailers)
+          } else {
+            let status = Status(
+              code: .unimplemented,
+              message: """
+                \(rawEncoding) compression is not supported; \
+                supported algorithms are listed in grpc-accept-encoding
+                """
+            )
+            var customTrailers = Metadata()
+            for acceptedEncoding in configuration.acceptedEncodings {
+              customTrailers.addString(
+                acceptedEncoding.name,
+                forKey: GRPCHTTP2Keys.acceptEncoding.rawValue
+              )
+            }
+            let trailers = self.makeTrailers(
+              status: status,
+              customMetadata: customTrailers,
+              trailersOnly: true
+            )
```

Comment:
Can we reduce the duplication here? The only diff is the status message and the custom metadata.